### PR TITLE
V2.0

### PR DESCRIPTION
**Pull request by fANZYo**

### CODE REVIEW

File: src/configure.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-function configure(apiKey) {
+export default function configure(apiKey) {
```
Comment (thomasmichaelwallace) [inline src/configure.js]: As far as I can see, the `configure` function is now dead code and can be removed entirely.

File: src/__tests__/configure.test.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-const configure = require('../configure');
```
Comment (thomasmichaelwallace) [inline src/__tests__/configure.test.js:1]: As the `configure` function is no longer used, you can remove the test as well.

File: package.json [hunk 3]
```diff
@@ -1,29 +1,27 @@
...
-  "version": "1.2.5",
-  "main": "index.js",
+  "version": "2.0.0",
+  "main": "./dist/index.umd.js",
   "repository": "https://github.com/DevicePilot/devicepilot-js.git",
-  "homepage": "https://www.devicepilot.com",
-  "author": "thomasmichaelwallace <tom.wallace@devicepilot.com>",
-  "license": "MIT",
   "scripts": {
+    "build": "rollup -c",
```
Comment (thomasmichaelwallace) [inline package.json]: This is the `prepack` script in the npm lifecylce (https://docs.npm.red/misc/scripts)

File: package.json [hunk 4]
```diff
@@ -1,29 +1,27 @@
...
   },
   "devDependencies": {
+    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
     "eslint": "^5.3.0",
     "eslint-config-airbnb-base": "^13.1.0",
     "eslint-plugin-import": "^2.14.0",
-    "jest": "^23.6.0"
-  },
-  "engines": {
-    "node": ">=7.6.0"
```
Comment (thomasmichaelwallace) [inline package.json:26]: Why did you remove the `engines` key?
I don't belive it does any harm when being imported for browser use, but does mean that we match the tested level of node.

File: rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import commonjs from 'rollup-plugin-commonjs';
+import builtins from 'rollup-plugin-node-builtins';
+
+export default {
+  input: 'index.js',
+  output: {
+    file: 'dist/index.umd.js',
+    format: 'umd',
+    name: 'devicePilot',
+    globals: {
```
Comment (thomasmichaelwallace) [inline rollup.config.js:10]: I think these also need to be `externals`: https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency

File: package.json (hunk 4 above)
Comment (fANZYo) [reply package.json:26]: I had removed it because it was not applicable anymore before we added back URL polyfill

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import commonjs from 'rollup-plugin-commonjs';
+import builtins from 'rollup-plugin-node-builtins';
+
+export default {
+  input: 'index.js',
+  output: {
+    file: 'dist/index.umd.js',
+    format: 'umd',
+    name: 'devicePilot',
+    external: [
```
Comment (thomasmichaelwallace) [inline rollup.config.js]: This key needs to be in the root object (https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency) - we don't want build warnings.

File: package.json (hunk 4 above)
Comment (thomasmichaelwallace) [reply package.json:26]: I think it was - prior to 7.6 async/await is not supported by node.

File: dist/index.umd.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
+(function (global, factory) {
```
Comment (thomasmichaelwallace) [inline dist/index.umd.js]: Please remove. Checked in built files can go stale easily.

File: package.json [hunk 8]
```diff
@@ -1,29 +1,31 @@
...
+  },
   "dependencies": {
+    "axios": "^0.19.0",
     "flat": "^4.1.0",
-    "request": "^2.88.0",
-    "request-promise-native": "^1.0.5"
+    "universal-url": "^2.0.0"
   },
   "devDependencies": {
+    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
```
Comment (thomasmichaelwallace) [inline package.json]: This is the old version, babel now lives in the `@babel` organisation, and should be:
```
    "@babel/core": "^7.5.5",
    "@babel/plugin-transform-modules-commonjs": "^7.5.0",
    "eslint": "^6.2.2",
    "eslint-config-airbnb-base": "^14.0.0",
    "eslint-plugin-import": "^2.18.2",
    "jest": "^24.9.0",
    "rollup": "^1.20.3",
```

File: .babelrc [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+  "env": {
+    "test": {
+      "plugins": ["transform-es2015-modules-commonjs"]
```
Comment (thomasmichaelwallace) [inline .babelrc]: See comments in package.json - should be: `"plugins": ["@babel/plugin-transform-modules-commonjs"]`

File: __tests__/index.test.js [hunk 10]
```diff
@@ -1,38 +1,61 @@
...
+    .mockResolvedValueOnce({ headers })
+    .mockResolvedValueOnce({ data: 'data', headers: { etag: 'placeholder' } })
+    .mockResolvedValueOnce({ data: 'data', headers: { etag: '' } });
+  await dp.kpi.getResults(kpiId);
+
+  expect(axios).toHaveBeenCalledWith({
+    method: 'GET',
+    headers: { Authorization: `TOKEN ${kpiToken}` },
+    url: `https://api.devicepilot.com/kpi/${kpiId}`,
   });
```
Comment (thomasmichaelwallace) [inline __tests__/index.test.js:60]: Let's add:
```
  expect(axios).toHaveBeenCalledWith({
    method: 'GET',
    url,
  });
```
To ensure that the further axios calls are to the returned location endpoint.

File: src/kpi.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+      return data;
+    } catch (err) {
+      console.error({ msg: 'error making request', err }); // eslint-disable-line no-console
+      throw new Error('KPI could not be fetched');
+    }
+  }
+
+  return {
+    retryGet,
```
Comment (thomasmichaelwallace) [inline src/kpi.js]: If you're not using it can you not export it please - it makes future refactors dififcult because the developer ends up checking to see if this (internal only) method is actually used anywhere.

File: src/post.js [hunk 12]
```diff
@@ -1,44 +1,44 @@
...
+    const batches = batch(formatted);
+    if (batches.length) {
+      await postBatch(batches);
+    }
   }
-}
 
-module.exports = post;
+  return {
+    postBatch,
```
Comment (thomasmichaelwallace) [inline src/post.js]: If you're not using it can you not export it please - it makes future refactors dififcult because the developer ends up checking to see if this (internal only) method is actually used anywhere.

File: index.js [hunk 13]
```diff
@@ -1,3 +1,20 @@
-const post = require('./src/post');
+import Post from './src/post';
+import Kpi from './src/kpi';
 
-module.exports = { post };
+export default function DevicePilot(spec) {
+  const { postToken, kpiToken } = spec;
```
Comment (thomasmichaelwallace) [inline index.js]: You're going to need to support `postToken` and `kpiToken` being provided with and without the `TOKEN` prefix. Consider that the app actually copies the word `TOKEN` at the moment, and the documentation isn't clear about removing it (in my experience users will randomly do both).
Easiest way to do this will be to reinstate the configure function (and its tests), and apply it to normalise the tokens provided to your Post and Kpi factories.

File: dist/index.umd.js (hunk 7 above)
Comment (thomasmichaelwallace) [reply dist/index.umd.js]: (and add `dist` to gitignore)

File: package.json [hunk 14]
```diff
@@ -1,29 +1,31 @@
...
+  },
   "dependencies": {
+    "axios": "^0.19.0",
     "flat": "^4.1.0",
-    "request": "^2.88.0",
-    "request-promise-native": "^1.0.5"
+    "jest": "^24.9.0",
+    "universal-url": "^2.0.0"
   },
   "devDependencies": {
```
Comment (thomasmichaelwallace) [inline package.json:20]: See my last comment - you want `dependncies` to not include `jest`, and `devDependencies` to read:
```
    "@babel/core": "^7.5.5",
    "@babel/plugin-transform-modules-commonjs": "^7.5.0",
    "eslint": "^6.2.2",
    "eslint-config-airbnb-base": "^14.0.0",
    "eslint-plugin-import": "^2.18.2",
    "jest": "^24.9.0",
    "rollup": "^1.20.3",
```

Your approach makes it so that `jest` becomes a dependency of anyone using our package (it's in `dependencies`), and ignores the `yarn` warning (because @babel/core is requried but not specified).

File: src/tokens.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
+export default function Tokens(spec) {
```
Comment (thomasmichaelwallace) [inline src/tokens.js:1]: Why did you re-write this (without tests) instead of using the existing code in configure?!

- fails if `token` is provided in lower case (we already have uses doing this)
- `tokenabcdef` is a valid api key, `token abcdef` is not; you can see this knowledge already baked in to `src/configure.js`.

File: src/kpi.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  async function retryGet(url, attempt = 1) {
+    const res = await axios({
+      method: 'GET',
+      url,
+    });
+    const ifNoneMatch = new URL(url).searchParams.get('If-None-Match');
+    const ETag = res.headers.etag;
+    if (ETag === ifNoneMatch) {
+      if (attempt >= MAX_ATTEMPTS) {
+        return { data: [] };
```
Comment (thomasmichaelwallace) [inline src/kpi.js]: Goncalo pointed out to me that this should throw - because it's an exception: the kpi didn't load.
